# Exploratory Data Dive: Titanic

In [1]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

# Load the dataset
df = sns.load_dataset("titanic")

In [2]:
# Display the first few rows of the dataset
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [3]:
# Display the last few rows of the dataset
df.tail()

survived  pclass     sex   age  sibsp  parch   fare embarked   class  \
886         0       2    male  27.0      0      0  13.00        S  Second   
887         1       1  female  19.0      0      0  30.00        S   First   
888         0       3  female   NaN      1      2  23.45        S   Third   
889         1       1    male  26.0      0      0  30.00        C   First   
890         0       3    male  32.0      0      0   7.75        Q   Third   

       who  adult_male deck  embark_town alive  alone  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True

In [4]:
# Display the shape of the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.7+ KB


In [5]:
# Display the columns of the dataset
df.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

# Column Types: Categorical vs Numeric

From the `.info()` output, we observe the dataset contains 15 columns. Here's the classification:

- **Categorical (non-numeric or encoded):**

  - `sex` (object): Passenger gender
  - `embarked` (object): Port of embarkation
  - `class` (category): Ticket class (First, Second, Third)
  - `who` (object): Passenger descriptor (man, woman, child)
  - `deck` (category): Deck level (A–G or NaN)
  - `embark_town` (object): Name of the port town
  - `alive` (object): Survival status
  - `alone` (bool): Whether the passenger was traveling alone
  - `adult_male` (bool): Boolean flag for adult male

- **Numeric:**

  - `survived` (int64): Binary indicator (0 = no, 1 = yes)
  - `pclass` (int64): Ticket class as integer
  - `age` (float64): Age in years
  - `sibsp` (int64): # of siblings/spouses aboard
  - `parch` (int64): # of parents/children aboard
  - `fare` (float64): Ticket fare

# Missing Data

Several columns have **missing values**, as evident from the `.info()` summary:

- `age`: 714 non-null → **177 missing**
- `deck`: 203 non-null → **688 missing**
- `embarked`: 889 non-null → **2 missing**
- `embark_town`: 825 non-null → **66 missing**

The `deck` column has the most significant proportion of missing data (~77%), which may limit its analytical utility unless imputed or treated separately.

# Notable Observations

- The `fare` column includes extreme values, ranging from **0.0** to **512.33**, suggesting outliers (likely 1st-class fares).
- The `age` field has a **broad spread** with a **mean of ~29.7** and a **standard deviation of ~14.5**, indicating a diverse passenger age distribution.


# 📊 Specific Feature Highlights

- **Range of Ages:**  

  - Minimum: **0.42 years**  
  - Maximum: **80.0 years**

- **Unique Values in `sex`:**  

  ```python
  df['sex'].unique()  # ['male', 'female']

In [6]:
# Display the number of values in the "sex" column
df["sex"].value_counts() 

sex
male      577
female    314
Name: count, dtype: int64

In [7]:
# Display the unique values in the "class" column
df["class"].unique()

['Third', 'First', 'Second']
Categories (3, object): ['First', 'Second', 'Third']

In [8]:
# Display the unique values in the "class" column and their counts
df.nunique()

survived         2
pclass           3
sex              2
age             88
sibsp            7
parch            7
fare           248
embarked         3
class            3
who              3
adult_male       2
deck             7
embark_town      3
alive            2
alone            2
dtype: int64

In [9]:
# Check the data types of the columns
print(df.dtypes)

# Explore unique values in non-numeric columns
print("\nUnique values in 'sex':", df['sex'].unique())
print("\nUnique values in 'embarked':", df['embarked'].unique())
print("\nUnique values in 'class':", df['class'].unique())

survived          int64
pclass            int64
sex              object
age             float64
sibsp             int64
parch             int64
fare            float64
embarked         object
class          category
who              object
adult_male         bool
deck           category
embark_town      object
alive            object
alone              bool
dtype: object

Unique values in 'sex': ['male' 'female']

Unique values in 'embarked': ['S' 'C' 'Q' nan]

Unique values in 'class': ['Third', 'First', 'Second']
Categories (3, object): ['First', 'Second', 'Third']


In [10]:
print("Missing values in 'age':", df["age"].isnull().sum())
print("Missing values in 'deck':", df["deck"].isnull().sum())
print("Missing values in 'embarked':", df["embarked"].isnull().sum())
print("Missing values in 'embark_town':", df["embark_town"].isnull().sum())

Missing values in 'age': 177
Missing values in 'deck': 688
Missing values in 'embarked': 2
Missing values in 'embark_town': 2


# Feature Relevance: Predictors & Grouping Candidates

# Candidate Predictor Variables

From an initial audit of the Titanic dataset, several variables stand out as promising predictors for modeling survival (`survived`):

- **`sex`**: A binary categorical variable (`male`, `female`) with a clear potential relationship to survival rates based on historical accounts.
- **`pclass`**: An ordinal indicator of ticket class (`1st`, `2nd`, `3rd`), which likely reflects socioeconomic status — often correlated with both fare and cabin location.
- **`age`**: A continuous variable with missing values, but potentially informative — especially if age distributions differ significantly between survivors and non-survivors.
- **`fare`**: A continuous variable that may serve as a proxy for wealth or accommodation quality.
- **`sibsp`** and **`parch`**: Count-based features indicating family structure onboard — potentially capturing whether a passenger was alone or with others, which could affect survival odds.
- **`embarked`**: A categorical port-of-boarding variable (`S`, `C`, `Q`) that may indirectly reflect passenger class, region, or boarding priority.

## Strong Grouping Variables for EDA

For initial data slicing and comparative analysis, the following categorical variables are well-suited for **grouping**:

- `sex`
- `pclass`
- `embarked`
- `who` (categorizes passengers as `man`, `woman`, `child`)
- `class` (redundant with `pclass` but in a cleaner categorical format)

These variables allow for aggregation and summary statistics (e.g., survival rates, mean age, average fare) and can support visual exploration through bar charts, box plots, or heatmaps.

## Less Useful (or Caution-Advised) Variables

- `deck`: Over 75% missing — high risk of bias if imputed without strong justification.
- `embark_town`: Redundant with `embarked`, and also contains missing values.
- `adult_male` and `alone`: Boolean engineered features that may overlap heavily with existing columns and reduce model interpretability if included naively.

## Summary

Variables like `sex`, `pclass`, `age`, and `fare` provide a strong blend of categorical and continuous features suitable for both grouping and prediction. These will form the foundation for hypothesis-driven exploration and any future modeling efforts. Special care must be taken in handling missing values, especially in `age` and `deck`, before using them for downstream tasks.

# Questions I want to explore

- Do women have a higher survival rate than men?
- Does fare or class correlate with survival?
- Are there missing age values? How might that affect analysis? 

In [11]:
# test